src/mlpack/methods/lmnn/lmnn_function_impl.hpp

```diff
@@ -351,37 +364,42 @@ double LMNNFunction<MetricType>::Evaluate(const arma::mat& transformation,
   // Apply metric over dataset.
   transformedDataset = transformation * dataset;
 
-  if (impBounds && iteration++ % range == 0)
+  if (recalculate)
   {
-    // Track number of data points to use for impostors calculatiom.
-    size_t numPoints = 0;
-
-    for (size_t i = begin; i < begin + batchSize; i++)
+    if (impBounds && iteration++ % range == 0)
     {
-      if (lastTransformationIndices(i))
+      // Track number of data points to use for impostors calculatiom.
+      size_t numPoints = 0;
+
+      for (size_t i = begin; i < begin + batchSize; i++)
       {
-        if (transformationDiffs[lastTransformationIndices[i]] *
-            (2 * norm(i) + norm(impostors(k - 1, i)) +
-            norm(impostors(k, i))) > distance(k, i) - distance(k - 1, i))
+        if (lastTransformationIndices(i))
         {
-          points(numPoints++)  = i;
+          if (transformationDiffs[lastTransformationIndices[i]] *
+              (2 * norm(i) + norm(impostors(k - 1, i)) +
+              norm(impostors(k, i))) > distance(k, i) - distance(k - 1, i))
+          {
+            points(numPoints++)  = i;
+          }
+        }
+        else
+        {
+          points(numPoints++) = i;
         }
       }
-      else
-      {
-        points(numPoints++) = i;
-      }
-    }
 
-    // Re-calculate impostors on transformed dataset.
-    constraint.Impostors(impostors, distance,
-        transformedDataset, labels, norm, points, numPoints);
-  }
-  else if (iteration++ % range == 0)
-  {
-    // Re-calculate impostors on transformed dataset.
-    constraint.Impostors(impostors, distance, transformedDataset, labels,
-        norm, begin, batchSize);
+      // Re-calculate impostors on transformed dataset.
+      constraint.Impostors(impostors, distance,
+          transformedDataset, labels, norm, points, numPoints);
```

Comment:
If I am understanding right, I think this is only calculating the impostors for the individual batch.  But we could calculate impostors for _all_ points in the dataset since it is the beginning of the epoch.

I think that to this end, we could move the impostor recalculation code to its own method (since I think it should be the same across all `Evaluate()`, `Gradient()`, and `EvaluateWithGradient()` overloads... correct me if I am wrong). And then in each of these functions we could have just a simple snippet:

```
if (recalculate)
{
  RecalculateImpostors(...); // this function sets 'recalculate' to false internally
}
```

What do you think, could that work?

Reply:
Actually `Shuffle()` is only called once, at the beginning of the epoch.  And typically it is also (slightly) faster to compute many impostors at once instead of doing them in small batches, due to the nature of the dual-tree nearest neighbor search algorithm.  Also if we compute them all at the beginning of the epoch, we can compute impostors with only one transformation matrix, which makes the handling of both the last transformation matrix easier and requires less computation to stretch the dataset (and the trees).  Also it makes the tree-related code I am working on a thousand times easier to write. ;)

So unless there's something I've overlooked, I think it would be a lot better to recompute all the impostors at once.  Let me know what you think.

Reply:
Then it's all good, earlier I thought that ``Shuffle()`` is being called during every batch. I will make the required changes.